src/AspNetCore/Inc/applicationinfo.h

```diff
@@ -128,19 +130,29 @@ class APPLICATION_INFO
         return m_pConfiguration;
     }
 
-    APPLICATION*
-    QueryApplication()
+    //
+    // ExtractApplication will increase the reference counter of the application
+    // Caller is responsible for dereference the application.
+    // Otherwise memory leak
+    // Open Issue: perf impact of the lock as this is in hot path per request
+    //
+    VOID
+    ExtractApplication(APPLICATION** ppApplication)
     {
-        return m_pApplication;
+        AcquireSRWLockExclusive(&m_srwLock);
```

Comment:
offline discussion: make this a shared lock as this doesn't modify the state of the application, only gets it.